lib/changelog/arc_ecto/schema.ex

```diff
@@ -0,0 +1,75 @@
+defmodule Changelog.Arc.Ecto.Schema do
+  defmacro __using__(_) do
+    quote do
+      import Changelog.Arc.Ecto.Schema
+    end
+  end
+
+  defmacro cast_attachments(changeset_or_data, params, allowed, options \\ []) do
+    quote bind_quoted: [changeset_or_data: changeset_or_data,
+                        params: params,
+                        allowed: allowed,
+                        options: options] do
+
+      # If given a changeset, apply the changes to obtain the underlying data
+      scope = case changeset_or_data do
+        %Ecto.Changeset{} -> Ecto.Changeset.apply_changes(changeset_or_data)
+        %{__meta__: _} -> changeset_or_data
+      end
+
+      # Cast supports both atom and string keys, ensure we're matching on both.
+      allowed = Enum.map(allowed, fn key ->
+        case key do
+          key when is_binary(key) -> key
+          key when is_atom(key) -> Atom.to_string(key)
+        end
+      end)
+
+      arc_params = case params do
+        :invalid ->
+          :invalid
+        %{} ->
+          params
+          |> Changelog.Arc.Ecto.Schema.convert_params_to_binary
+          |> Map.take(allowed)
+          |> Enum.reduce([], fn
+            # Don't wrap nil casts in the scope object
+            {field, nil}, fields -> [{field, nil} | fields]
+
+            # Allow casting Plug.Uploads
+            {field, upload = %{__struct__: Plug.Upload}}, fields -> [{field, {upload, scope}} | fields]
+
+            # If casting a binary (path), ensure we've explicitly allowed paths or urls
+            {field, path}, fields when is_binary(path) ->
+              if Keyword.get(options, :allow_urls, false) and Regex.match?( ~r/^https?:\/\// , path) do
```

Comment:
Instead of a nested `if`, this might read better as a `cond` statement:

```elixir
cond do
  Keyword.get(options, :allow_urls, false) and Regex.match?( ~r/^https?:\/\// , path) -> [{field, {path, scope}} | fields]
  Keyword.get(options, :allow_paths, false) -> [{field, {path, scope}} | fields]
  true -> fields
end
```